stl/inc/deque

```diff
@@ -986,19 +986,63 @@ public:
     }
 
     void shrink_to_fit() {
-        size_type _Oldcapacity = _Block_size * _Mapsize();
-        size_type _Newcapacity = _Oldcapacity / 2;
+        if (empty()) {
+            if (_Map() != nullptr) {
+                _Reset_map();
+            }
+            return;
+        }
 
-        if (_Newcapacity < _Block_size * _Minimum_map_size) {
-            _Newcapacity = _Block_size * _Minimum_map_size;
+        const auto _First_block_idx = static_cast<_Map_difference_type>(_Myoff() / _Block_size);
+        const auto _Last_block_idx =
+            static_cast<_Map_difference_type>(((_Myoff() + _Mysize() - 1) / _Block_size + 1) & (_Mapsize() - 1));
+
+        // deallocate unused blocks
+        for (auto _Block_idx = _Last_block_idx; _Block_idx != _First_block_idx;
+             _Block_idx      = static_cast<_Map_difference_type>((_Block_idx + 1) & (_Mapsize() - 1))) {
+            auto& _Block_ptr = _Map()[_Block_idx];
+            if (_Block_ptr != nullptr) {
+                _Getal().deallocate(_Block_ptr, _Block_size);
+                _Block_ptr = nullptr;
+            }
         }
 
-        if ((empty() && _Mapsize() > 0)
-            || (!empty() && size() <= _Newcapacity && _Newcapacity < _Oldcapacity)) { // worth shrinking, do it
-            deque _Tmp(
-                _STD make_move_iterator(_Unchecked_begin()), _STD make_move_iterator(_Unchecked_end()), _Getal());
-            swap(_Tmp);
+        const auto _Used_block_count =
+            static_cast<size_type>((_Myoff() + _Mysize() - 1) / _Block_size + 1 - _Myoff() / _Block_size);
+
+        size_type _New_block_count = _Minimum_map_size; // should be power of 2
+        for (; _New_block_count < _Used_block_count; _New_block_count *= 2)
+            ;
+        if (_New_block_count >= _Mapsize()) {
+            return;
         }
+
+        // worth shrinking the internal map, do it
+
+        _Alpty _Almap(_Getal());
+        const auto _New_map = _Almap.allocate(_New_block_count);
+
+        _Orphan_all(); // the map can be shrinked, invalidate all iterators
+
+        _Map_difference_type _New_block_idx = 0;
+        for (auto _Block_idx = _First_block_idx; _Block_idx != _Last_block_idx;
+             _Block_idx = static_cast<_Map_difference_type>((_Block_idx + 1) & (_Mapsize() - 1)), ++_New_block_idx) {
+            // transfer the ownership of blocks to pointers in the new map
+            _STD _Construct_in_place(_New_map[_New_block_idx], _Map()[_Block_idx]);
+        }
+        // zero out the rest of the new map
+        _STD _Uninitialized_value_construct_n_unchecked1(
+            _New_map + _New_block_idx, _New_block_count - static_cast<size_type>(_New_block_idx));
+
+        for (auto _Block = _Map_distance(); _Block > 0;) {

```

Comment:
Should we add if( _Map() != nullptr) before deallocating map?

Reply:
I believe we shouldn't. At the point where `deallocate` gets called, the `deque` must be non-empty, so `_Map()` must be non-null.